File: archiso/airootfs/usr/share/churros/preferences/pages/window_rules.py
import sys
import os

ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
PREF = os.path.abspath(os.path.join(ROOT, "preferences"))

sys.path.insert(0, ROOT)
sys.path.insert(0, PREF)

import gi

gi.require_version("Gtk", "4.0")

from gi.repository import Gtk, GLib

from widgets.page import Page
from widgets.group import Group
from widgets.row import Row
from widgets.switch_row import SwitchRow

from services.window_rules_service import WindowRulesService
from services.dotfiles.niri_config import NiriConfig


class WindowRulesPage(Page):

    def __init__(self, navigator):

        super().__init__(
            navigator,
            "Reglas de ventana",
            "Window-rules de Niri: opacidad, flotantes, esquinas, blur",
            parent_page="appearance"
        )

        self._pending = []

        self._build()

    def _build(self):

        #
        # Lista de reglas
        #

        self._rules_group = Group("Reglas definidas")

        self._rules_list = Gtk.Box(
            orientation=Gtk.Orientation.VERTICAL,
            spacing=8
        )

        self._rules_group.add(self._rules_list)

        self.add(self._rules_group)

        self._refresh_rules()

        #
        # Anadir nueva regla
        #

        add_group = Group("Anadir regla")

        self._app_id_entry = Gtk.Entry()
        self._app_id_entry.set_placeholder_text("app-id, p.ej. firefox")
        self._app_id_entry.set_margin_start(14)
        self._app_id_entry.set_margin_end(14)
        self._app_id_entry.set_margin_top(10)
        self._app_id_entry.set_margin_bottom(10)

        add_group.add(self._app_id_entry)

        self._title_entry = Gtk.Entry()
        self._title_entry.set_placeholder_text('regex del titulo (opcional)')
        self._title_entry.set_margin_start(14)
        self._title_entry.set_margin_end(14)
        self._title_entry.set_margin_bottom(10)

        add_group.add(self._title_entry)

        opacity_box = Gtk.Box(
            orientation=Gtk.Orientation.HORIZONTAL,
            spacing=8
        )
        opacity_box.set_margin_start(14)
        opacity_box.set_margin_end(14)
        opacity_box.set_margin_bottom(10)

        opacity_label = Gtk.Label(label="Opacidad:")
        opacity_label.set_xalign(0)
        opacity_label.set_hexpand(True)

        self._opacity_spin = Gtk.SpinButton.new_with_range(
            0.0, 1.0, 0.05
        )
        self._opacity_spin.set_value(1.0)

        opacity_box.append(opacity_label)
        opacity_box.append(self._opacity_spin)

        add_group.add(opacity_box)

        corner_box = Gtk.Box(
            orientation=Gtk.Orientation.HORIZONTAL,
            spacing=8
        )
        corner_box.set_margin_start(14)
        corner_box.set_margin_end(14)
        corner_box.set_margin_bottom(10)

        corner_label = Gtk.Label(label="Radio de esquinas:")
        corner_label.set_xalign(0)
        corner_label.set_hexpand(True)

        self._corner_spin = Gtk.SpinButton.new_with_range(
            0.0, 64.0, 1.0
        )
        self._corner_spin.set_value(0.0)

        corner_box.append(corner_label)
        corner_box.append(self._corner_spin)

        add_group.add(corner_box)

        self._floating_switch = SwitchRow(
            title="Abrir como flotante",
            subtitle="Las ventanas que matcheen se abren como popups",
            active=False,
            callback=lambda *_: None
        )

        add_group.add(self._floating_switch)

        self._clip_switch = SwitchRow(
            title="Recortar al geometry-corner-radius",
            subtitle="clip-to-geometry",
            active=False,
            callback=lambda *_: None
        )

        add_group.add(self._clip_switch)

        self._blur_switch = SwitchRow(
            title="Fondo con blur (background-effect)",
            subtitle="Desenfoque del fondo detras de la ventana",
            active=False,
            callback=lambda *_: None
        )

        add_group.add(self._blur_switch)

        add_row = Row(
            title="Anadir regla",
            subtitle="Crea una nueva window-rule con estos valores",
            icon="window_rules.svg",
            value=None
        )

        add_row.connect(
            "clicked",
            lambda *_: self._on_add_rule()
        )

        add_group.add(add_row)

        self._edit_index = None

        self._update_row = Row(
            title="Guardar cambios sobre la regla seleccionada",
            subtitle="Actualizar la regla que estas editando",
            icon="window_rules.svg",
            value=None
        )

        self._update_row.connect(
            "clicked",
            lambda *_: self._on_update_rule()
        )

        self._update_row.set_visible(False)

        add_group.add(self._update_row)

        self.add(add_group)

        #
        # Acciones
        #

        actions_group = Group("Acciones")

        reload_row = Row(
            title="Recargar Niri",
            subtitle="Aplica los cambios forzando una transicion",
            icon="logs.svg",
            value=None
        )

        reload_row.connect(
            "clicked",
            lambda *_: NiriConfig.reload()
        )

        actions_group.add(reload_row)

        self.add(actions_group)

    def _refresh_rules(self):

        child = self._rules_list.get_first_child()

        while child is not None:

            nxt = child.get_next_sibling()
            self._rules_list.remove(child)
            child = nxt

        try:

            rules = WindowRulesService.list_rules()

        except Exception as exc:

            label = Gtk.Label(label="Error: " + str(exc))
            label.set_xalign(0)
            label.set_wrap(True)
            self._rules_list.append(label)

            return

        if not rules:

            label = Gtk.Label(label="No hay reglas definidas.")
            label.set_xalign(0)
            label.add_css_class("row-subtitle")
            label.set_margin_start(14)
            label.set_margin_top(10)
            label.set_margin_bottom(10)

            self._rules_list.append(label)

            return

        for r in rules:

            self._rules_list.append(
                self._build_rule_card(r)
            )

    def _build_rule_card(
        self,
        rule
    ):

        card = Gtk.Box(
            orientation=Gtk.Orientation.VERTICAL,
            spacing=6
        )
        card.add_css_class("row")
        card.set_margin_top(10)
        card.set_margin_bottom(10)
        card.set_margin_start(14)
        card.set_margin_end(14)

        title = rule["app_id"] or rule["title"] or "(sin filtro)"
        title_label = Gtk.Label(label="Regla: " + title)
        title_label.set_xalign(0)
        title_label.set_hexpand(True)
        title_label.add_css_class("row-title")

        card.append(title_label)

        summary_parts = []

        if rule.get("title") and rule.get("app_id"):
            summary_parts.append('title="' + rule["title"] + '"')

        if rule.get("opacity") is not None:
            summary_parts.append("opacity " + str(rule["opacity"]))

        if rule.get("open_floating") is not None:
            summary_parts.append(
                "open-floating " +
                ("true" if rule["open_floating"] else "false")
            )

        if rule.get("corner_radius") is not None:
            summary_parts.append(
                "radius " + str(rule["corner_radius"])
            )

        if rule.get("clip_to_geometry") is not None:
            summary_parts.append(
                "clip " +
                ("true" if rule["clip_to_geometry"] else "false")
            )

        if rule.get("blur") is not None:
            summary_parts.append(
                "blur " + ("true" if rule["blur"] else "false")
            )

        summary = Gtk.Label(
            label="    " + (", ".join(summary_parts) or "(sin cambios)")
        )
        summary.set_xalign(0)
        summary.set_hexpand(True)
        summary.add_css_class("row-subtitle")
        summary.set_wrap(True)

        card.append(summary)

        actions = Gtk.Box(
            orientation=Gtk.Orientation.HORIZONTAL,
            spacing=8
        )
        actions.set_halign(Gtk.Align.END)

        edit_btn = Gtk.Button(label="Editar")
        edit_btn.add_css_class("suggested-action")
        edit_btn.connect(
            "clicked",
            lambda *_: self._on_edit_rule(rule["index"])
        )

        del_btn = Gtk.Button(label="Borrar")
        del_btn.add_css_class("destructive-action")
        del_btn.connect(
            "clicked",
            lambda *_: self._on_delete_rule(rule["index"])
        )

        actions.append(edit_btn)
        actions.append(del_btn)

        card.append(actions)

        return card

    def _on_add_rule(self):

        app_id = self._app_id_entry.get_text().strip()
        title = self._title_entry.get_text().strip()

        opacity = self._opacity_spin.get_value()
        if opacity >= 1.0:
            opacity = None

        corner_radius = self._corner_spin.get_value()
        if corner_radius == 0.0:
            corner_radius = None

        try:

            WindowRulesService.add_rule(
                app_id=app_id,
                title=title,
                opacity=opacity,
                open_floating=self._floating_switch.get_active() or None,
                corner_radius=corner_radius,
                clip_to_geometry=self._clip_switch.get_active() or None,
                blur=self._blur_switch.get_active() or None,
            )

            self._reset_form()

            NiriConfig.reload()

            self._refresh_rules()

        except Exception as exc:

            dlg = Gtk.AlertDialog()
            dlg.set_heading("Error")
            dlg.set_message(str(exc))
            dlg.show(self.get_root())

    def _on_update_rule(self):

        if self._edit_index is None:
            return

        app_id = self._app_id_entry.get_text().strip()
        title = self._title_entry.get_text().strip()

        update = {
            "app_id": app_id,
            "title": title,
        }

        opacity = self._opacity_spin.get_value()
        update["opacity"] = None if opacity >= 1.0 else opacity

        corner_radius = self._corner_spin.get_value()
        update["corner_radius"] = None if corner_radius == 0.0 \
            else corner_radius

        update["open_floating"] = \
            self._floating_switch.get_active() or None

        update["clip_to_geometry"] = \
            self._clip_switch.get_active() or None

        update["blur"] = self._blur_switch.get_active() or None

        try:

            WindowRulesService.update_rule(self._edit_index, **update)

            self._edit_index = None

            self._update_row.set_visible(False)

            self._reset_form()

            NiriConfig.reload()

            self._refresh_rules()

        except Exception as exc:

            dlg = Gtk.AlertDialog()
            dlg.set_heading("Error")
            dlg.set_message(str(exc))
            dlg.show(self.get_root())

    def _reset_form(self):

        self._app_id_entry.set_text("")
        self._title_entry.set_text("")
        self._opacity_spin.set_value(1.0)
        self._corner_spin.set_value(0.0)
        self._floating_switch.set_active(False)
        self._clip_switch.set_active(False)
        self._blur_switch.set_active(False)

    def _on_delete_rule(self, index):

        dlg = Gtk.AlertDialog()
        dlg.set_heading("Borrar regla")
        dlg.set_message("¿Seguro que quieres borrar esta regla?")
        dlg.set_modal(True)
        dlg.set_buttons(["Cancelar", "Borrar"])

        def on_resp(d, r):

            try:
                resp = d.choose_finish(r)
            except Exception:
                return

            if resp != 1:
                return

            try:

                WindowRulesService.delete_rule(index)

                NiriConfig.reload()

                self._refresh_rules()

            except Exception as exc:

                dlg2 = Gtk.AlertDialog()
                dlg2.set_heading("Error")
                dlg2.set_message(str(exc))
                dlg2.show(self.get_root())

        dlg.choose(self.get_root(), None, on_resp)

    def _on_edit_rule(self, index):

        try:
            rules = WindowRulesService.list_rules()
        except Exception as exc:
            return

        for r in rules:
            if r["index"] == index:
                match = r
                break
        else:
            return

        self._app_id_entry.set_text(match.get("app_id") or "")
        self._title_entry.set_text(match.get("title") or "")

        if match.get("opacity") is not None:
            self._opacity_spin.set_value(float(match["opacity"]))

        if match.get("corner_radius") is not None:
            self._corner_spin.set_value(float(match["corner_radius"]))

        self._floating_switch.set_active(
            bool(match.get("open_floating", False))
        )
        self._clip_switch.set_active(
            bool(match.get("clip_to_geometry", False))
        )
        self._blur_switch.set_active(
            bool(match.get("blur", False))
        )

        self._edit_index = index
        self._update_row.set_visible(True)

File: rust/preferences/src/pages/window_rules.rs
// ==========================================
// WindowRulesPage — window-rules de Niri
// (equivalente a pages/window_rules.py)
// ==========================================

use gtk::prelude::*;

use std::cell::RefCell;
use std::rc::Rc;

use crate::services::niri_config::NiriConfig;
use crate::services::window_rules_service::WindowRulesService;
use crate::widgets::group::Group;
use crate::widgets::page::Page;
use crate::widgets::row::Row;
use crate::widgets::switch_row::SwitchRow;

// AsWidget para widgets GTK crudos añadidos a un Group
// (el Group::add acepta &impl AsWidget)
struct WBox(gtk::Box);
impl crate::widgets::AsWidget for WBox {
    fn widget(&self) -> &gtk::Widget {
        self.0.upcast_ref()
    }
}
struct WEntry(gtk::Entry);
impl crate::widgets::AsWidget for WEntry {
    fn widget(&self) -> &gtk::Widget {
        self.0.upcast_ref()
    }
}
struct WSpin(gtk::SpinButton);
impl crate::widgets::AsWidget for WSpin {
    fn widget(&self) -> &gtk::Widget {
        self.0.upcast_ref()
    }
}
struct WLabel(gtk::Label);
impl crate::widgets::AsWidget for WLabel {
    fn widget(&self) -> &gtk::Widget {
        self.0.upcast_ref()
    }
}

struct RulesState {
    app_id_entry: gtk::Entry,
    title_entry: gtk::Entry,
    opacity_spin: gtk::SpinButton,
    corner_spin: gtk::SpinButton,
    floating_switch: SwitchRow,
    clip_switch: SwitchRow,
    blur_switch: SwitchRow,
    edit_index: Option<usize>,
    update_row: Option<Row>,
    rules_list: gtk::Box,
}

type RulesStateRef = Rc<RefCell<RulesState>>;

pub fn build(navigator: gtk::Stack) -> Page {
    let page = Page::new(
        Some(navigator.clone()),
        "Reglas de ventana",
        Some("Window-rules de Niri: opacidad, flotantes, esquinas, blur"),
        Some("appearance".to_string()),
    );

    // ---------- Lista de reglas ----------
    let mut rules_group = Group::new("Reglas definidas");

    let rules_list = gtk::Box::new(gtk::Orientation::Vertical, 8);
    rules_group.add(&WBox(rules_list.clone()));

    page.add(rules_group.widget());

    // ---------- Anadir regla ----------
    let mut add_group = Group::new("Anadir regla");

    let app_id_entry = gtk::Entry::new();
    app_id_entry.set_placeholder_text(Some("app-id, p.ej. firefox"));
    app_id_entry.set_margin_start(14);
    app_id_entry.set_margin_end(14);
    app_id_entry.set_margin_top(10);
    app_id_entry.set_margin_bottom(10);
    add_group.add(&WEntry(app_id_entry.clone()));

    let title_entry = gtk::Entry::new();
    title_entry.set_placeholder_text(Some("regex del titulo (opcional)"));
    title_entry.set_margin_start(14);
    title_entry.set_margin_end(14);
    title_entry.set_margin_bottom(10);
    add_group.add(&WEntry(title_entry.clone()));

    // Opacidad
    let opacity_box = gtk::Box::new(gtk::Orientation::Horizontal, 8);
    opacity_box.set_margin_start(14);
    opacity_box.set_margin_end(14);
    opacity_box.set_margin_bottom(10);

    let opacity_label = gtk::Label::new(Some("Opacidad:"));
    opacity_label.set_xalign(0.0);
    opacity_label.set_hexpand(true);

    let opacity_spin = gtk::SpinButton::with_range(0.0, 1.0, 0.05);
    opacity_spin.set_value(1.0);

    opacity_box.append(&opacity_label);
    opacity_box.append(&opacity_spin);
    add_group.add(&WBox(opacity_box));

    // Radio de esquinas
    let corner_box = gtk::Box::new(gtk::Orientation::Horizontal, 8);
    corner_box.set_margin_start(14);
    corner_box.set_margin_end(14);
    corner_box.set_margin_bottom(10);

    let corner_label = gtk::Label::new(Some("Radio de esquinas:"));
    corner_label.set_xalign(0.0);
    corner_label.set_hexpand(true);

    let corner_spin = gtk::SpinButton::with_range(0.0, 64.0, 1.0);
    corner_spin.set_value(0.0);

    corner_box.append(&corner_label);
    corner_box.append(&corner_spin);
    add_group.add(&WBox(corner_box));

    let floating_switch = SwitchRow::new(
        "Abrir como flotante",
        None,
        Some("Las ventanas que matcheen se abren como popups"),
        false,
        Some(Box::new(|_| {})),
    );
    add_group.add(&floating_switch);

    let clip_switch = SwitchRow::new(
        "Recortar al geometry-corner-radius",
        None,
        Some("clip-to-geometry"),
        false,
        Some(Box::new(|_| {})),
    );
    add_group.add(&clip_switch);

    let blur_switch = SwitchRow::new(
        "Fondo con blur (background-effect)",
        None,
        Some("Desenfoque del fondo detras de la ventana"),
        false,
        Some(Box::new(|_| {})),
    );
    add_group.add(&blur_switch);

    // Fila "Anadir regla"
    let state: RulesStateRef = Rc::new(RefCell::new(RulesState {
        app_id_entry: app_id_entry.clone(),
        title_entry: title_entry.clone(),
        opacity_spin: opacity_spin.clone(),
        corner_spin: corner_spin.clone(),
        floating_switch,
        clip_switch,
        blur_switch,
        edit_index: None,
        update_row: None,
        rules_list: rules_list.clone(),
    }));

    {
        let st = Rc::clone(&state);
        add_group.add(&Row::new(
            "Anadir regla",
            Some("Crea una nueva window-rule con estos valores"),
            Some("window_rules.svg"),
            None,
            None,
            Some(Box::new(move |_| on_add_rule(&st))),
        ));
    }

    // Fila "Guardar cambios" (oculta al inicio)
    let update_row = Row::new(
        "Guardar cambios sobre la regla seleccionada",
        Some("Actualizar la regla que estas editando"),
        Some("window_rules.svg"),
        None,
        None,
        None,
    );
    update_row.widget().set_visible(false);
    {
        let st = Rc::clone(&state);
        update_row
            .widget()
            .connect_clicked(move |_| on_update_rule(&st));
    }
    state.borrow_mut().update_row = Some(update_row);
    {
        let st = state.borrow();
        add_group.add(st.update_row.as_ref().unwrap());
    }

    page.add(add_group.widget());

    // ---------- Acciones ----------
    let mut actions_group = Group::new("Acciones");

    actions_group.add(&Row::new(
        "Recargar Niri",
        Some("Aplica los cambios forzando una transicion"),
        Some("logs.svg"),
        None,
        None,
        Some(Box::new(|_| NiriConfig::reload())),
    ));

    page.add(actions_group.widget());

    // Cargar reglas
    refresh_rules(&state);

    page
}

/// Vacía la lista y la rellena con las reglas (equivalente a _refresh_rules).
fn refresh_rules(state: &RulesStateRef) {
    let st = state.borrow();

    while let Some(child) = st.rules_list.first_child() {
        st.rules_list.remove(&child);
    }

    let rules = WindowRulesService::list_rules();
    if rules.is_empty() {
        let label = gtk::Label::new(Some("Sin reglas o error al leer config.kdl"));
        label.set_xalign(0.0);
        label.set_wrap(true);
        st.rules_list.append(&label);
        return;
    }

    if rules.is_empty() {
        let label = gtk::Label::new(Some("No hay reglas definidas."));
        label.set_xalign(0.0);
        label.add_css_class("row-subtitle");
        label.set_margin_start(14);
        label.set_margin_top(10);
        label.set_margin_bottom(10);
        st.rules_list.append(&label);
        return;
    }

    for rule in &rules {
        st.rules_list.append(&build_rule_card(state, rule));
    }
}

/// Tarjeta de regla con Editar/Borrar (equivalente a _build_rule_card).
fn build_rule_card(state: &RulesStateRef, rule: &serde_json::Value) -> gtk::Box {
    let card = gtk::Box::new(gtk::Orientation::Vertical, 6);
    card.add_css_class("row");
    card.set_margin_top(10);
    card.set_margin_bottom(10);
    card.set_margin_start(14);
    card.set_margin_end(14);

    let app_id = rule["app_id"].as_str().unwrap_or("");
    let title = rule["title"].as_str().unwrap_or("");
    let card_title = if !app_id.is_empty() || !title.is_empty() {
        if !app_id.is_empty() { app_id } else { title }
    } else {
        "(sin filtro)"
    };

    let title_label = gtk::Label::new(Some(&format!("Regla: {card_title}")));
    title_label.set_xalign(0.0);
    title_label.set_hexpand(true);
    title_label.add_css_class("row-title");
    card.append(&title_label);

    // Resumen
    let mut summary_parts: Vec<String> = Vec::new();

    if !title.is_empty() && !app_id.is_empty() {
        summary_parts.push(format!("title=\"{title}\""));
    }
    if let Some(opacity) = rule["opacity"].as_f64() {
        summary_parts.push(format!("opacity {opacity}"));
    }
    if let Some(open_floating) = rule["open_floating"].as_bool() {
        summary_parts.push(format!("open-floating {open_floating}"));
    }
    if let Some(corner_radius) = rule["corner_radius"].as_f64() {
        summary_parts.push(format!("radius {corner_radius}"));
    }
    if let Some(clip) = rule["clip_to_geometry"].as_bool() {
        summary_parts.push(format!("clip {clip}"));
    }
    if let Some(blur) = rule["blur"].as_bool() {
        summary_parts.push(format!("blur {blur}"));
    }

    let summary_text = if summary_parts.is_empty() {
        "(sin cambios)".to_string()
    } else {
        summary_parts.join(", ")
    };

    let summary = gtk::Label::new(Some(&format!("    {summary_text}")));
    summary.set_xalign(0.0);
    summary.set_hexpand(true);
    summary.add_css_class("row-subtitle");
    summary.set_wrap(true);
    card.append(&summary);

    // Acciones
    let actions = gtk::Box::new(gtk::Orientation::Horizontal, 8);
    actions.set_halign(gtk::Align::End);

    let index = rule["index"].as_u64().unwrap_or(0) as usize;

    let edit_btn = gtk::Button::with_label("Editar");
    edit_btn.add_css_class("suggested-action");
    let st = Rc::clone(state);
    edit_btn.connect_clicked(move |_| on_edit_rule(&st, index));

    let del_btn = gtk::Button::with_label("Borrar");
    del_btn.add_css_class("destructive-action");
    let st = Rc::clone(state);
    del_btn.connect_clicked(move |btn| on_delete_rule(btn, &st, index));

    actions.append(&edit_btn);
    actions.append(&del_btn);
    card.append(&actions);

    card
}

/// Equivalente a _on_add_rule.
fn on_add_rule(state: &RulesStateRef) {
    let st = state.borrow();
    let app_id = st.app_id_entry.text().trim().to_string();
    let title = st.title_entry.text().trim().to_string();

    let opacity = st.opacity_spin.value();
    let opacity = if opacity >= 1.0 { None } else { Some(opacity) };

    let corner_radius = st.corner_spin.value();
    let corner_radius = if corner_radius == 0.0 { None } else { Some(corner_radius) };

    let open_floating = st.floating_switch.get_active().then_some(true);
    let clip_to_geometry = st.clip_switch.get_active().then_some(true);
    let blur = st.blur_switch.get_active().then_some(true);
    drop(st);

    match WindowRulesService::add_rule(
        &app_id,
        &title,
        opacity,
        open_floating,
        corner_radius,
        clip_to_geometry,
        blur,
    ) {
        _ => {}
    }

    reset_form(state);
    NiriConfig::reload();
    refresh_rules(state);
}

/// Equivalente a _on_update_rule.
fn on_update_rule(state: &RulesStateRef) {
    let st = state.borrow();
    let Some(index) = st.edit_index else {
        return;
    };

    let app_id = st.app_id_entry.text().trim().to_string();
    let title = st.title_entry.text().trim().to_string();

    let opacity = st.opacity_spin.value();
    let opacity = if opacity >= 1.0 { Value::Null } else { serde_json::json!(opacity) };

    let corner_radius = st.corner_spin.value();
    let corner_radius = if corner_radius == 0.0 { Value::Null } else { serde_json::json!(corner_radius) };

    let updates = serde_json::json!({
        "app_id": app_id,
        "title": title,
        "opacity": opacity,
        "corner_radius": corner_radius,
        // Python: get_active() or None -> None si False
        "open_floating": st.floating_switch.get_active().then_some(true),
        "clip_to_geometry": st.clip_switch.get_active().then_some(true),
        "blur": st.blur_switch.get_active().then_some(true),
    });
    drop(st);

    match WindowRulesService::update_rule(index, &updates) {
        Ok(()) => {
            {
                let mut st = state.borrow_mut();
                st.edit_index = None;
                if let Some(row) = st.update_row.as_ref() {
                    row.widget().set_visible(false);
                }
            }
            reset_form(state);
            NiriConfig::reload();
            refresh_rules(state);
        }
        Err(e) => show_error(state, &e),
    }
}

use serde_json::Value;

/// Equivalente a _reset_form.
fn reset_form(state: &RulesStateRef) {
    let st = state.borrow();
    st.app_id_entry.set_text("");
    st.title_entry.set_text("");
    st.opacity_spin.set_value(1.0);
    st.corner_spin.set_value(0.0);
    st.floating_switch.set_active(false);
    st.clip_switch.set_active(false);
    st.blur_switch.set_active(false);
}

/// Equivalente a _on_delete_rule (con confirmacion).
fn on_delete_rule(btn: &gtk::Button, state: &RulesStateRef, index: usize) {
    let dialog = gtk::AlertDialog::builder()
        .message("¿Seguro que quieres borrar esta regla?")
        .modal(true)
        .buttons(["Cancelar", "Borrar"])
        .build();

    let window = btn.root().and_downcast::<gtk::Window>();
    let st = Rc::clone(state);

    dialog.choose(window.as_ref(), None::<&gio::Cancellable>, move |result| {
        let Ok(response) = result else {
            return;
        };
        if response != 1 {
            return;
        }

        match WindowRulesService::delete_rule(index) {
            Ok(()) => {
                NiriConfig::reload();
                refresh_rules(&st);
            }
            Err(e) => show_error(&st, &e),
        }
    });
}

/// Equivalente a _on_edit_rule: rellena el formulario con la regla.
fn on_edit_rule(state: &RulesStateRef, index: usize) {
    let rules = WindowRulesService::list_rules();
    if rules.is_empty() {
        return;
    }

    let Some(match_rule) = rules.iter().find(|r| r["index"].as_u64() == Some(index as u64)) else {
        return;
    };

    let st = state.borrow();
    st.app_id_entry.set_text(match_rule["app_id"].as_str().unwrap_or(""));
    st.title_entry.set_text(match_rule["title"].as_str().unwrap_or(""));

    if let Some(opacity) = match_rule["opacity"].as_f64() {
        st.opacity_spin.set_value(opacity);
    }
    if let Some(corner_radius) = match_rule["corner_radius"].as_f64() {
        st.corner_spin.set_value(corner_radius);
    }
    st.floating_switch
        .set_active(match_rule["open_floating"].as_bool().unwrap_or(false));
    st.clip_switch
        .set_active(match_rule["clip_to_geometry"].as_bool().unwrap_or(false));
    st.blur_switch
        .set_active(match_rule["blur"].as_bool().unwrap_or(false));
    drop(st);

    state.borrow_mut().edit_index = Some(index);
    if let Some(row) = state.borrow().update_row.as_ref() {
        row.widget().set_visible(true);
    }
}

/// Dialogo de error (equivalente al Gtk.AlertDialog del Python).
fn show_error(state: &RulesStateRef, message: &str) {
    let dialog = gtk::AlertDialog::builder()
        .message(message)
        .build();

    // Necesitamos un parent: usamos la raiz de cualquier widget del estado
    let root = state
        .borrow()
        .rules_list
        .root()
        .and_downcast::<gtk::Window>();
    dialog.show(root.as_ref());
}
